pkg/admin.go

```diff
@@ -1,20 +1,76 @@
 package pkg
 
 import (
-	"github.com/nuts-foundation/nuts-crypto/pkg"
+	"crypto/x509"
+	"encoding/base64"
+	"errors"
+	"github.com/lestrrat-go/jwx/jwk"
+	crypto "github.com/nuts-foundation/nuts-crypto/pkg"
 	"github.com/nuts-foundation/nuts-crypto/pkg/types"
+	"github.com/nuts-foundation/nuts-registry/pkg/cert"
 	"github.com/nuts-foundation/nuts-registry/pkg/events"
+	errors2 "github.com/pkg/errors"
 	"github.com/sirupsen/logrus"
 	"time"
 )
 
+// ErrJWKConstruction indicates that a JWK couldn't be constructed
+var ErrJWKConstruction = errors.New("unable to construct JWK")
+
 // RegisterVendor registers a vendor
-func (r *Registry) RegisterVendor(id string, name string) (events.Event, error) {
-	logrus.Infof("Registering vendor, id=%s, name=%s", id, name)
-	return r.publishEvent(events.RegisterVendor, events.RegisterVendorEvent{
+func (r *Registry) RegisterVendor(id string, name string, domain string) (events.Event, error) {
+	r.logger().Infof("Registering vendor, id=%s, name=%s, domain=%s", id, name, domain)
+	entity := types.LegalEntity{URI: id}
+	err := r.crypto.GenerateKeyPairFor(entity)
+	if err != nil {
+		return nil, err
+	}
+
+	certASN1, err := cert.IssueCertificate(r.crypto, func() (x509.CertificateRequest, error) {
+		// TODO: Make env configurable
+		return cert.VendorCACertificateRequest(id, name, domain, "")
+	}, entity, entity, crypto.CertificateProfile{
+		KeyUsage:     x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign | x509.KeyUsageCRLSign,
+		NumDaysValid: r.Config.VendorCACertificateValidity,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	jwkAsMap, err := marshalJwk(certASN1, cert.VendorCACertificate)
+	if err != nil {
+		return nil, errors2.Wrap(err, ErrJWKConstruction.Error())
+	}
+
+	event := events.CreateEvent(events.RegisterVendor, events.RegisterVendorEvent{
 		Identifier: events.Identifier(id),
 		Name:       name,
+		Domain:     domain,
+		Keys:       []interface{}{jwkAsMap},
 	})
+	err = r.EventSystem.PublishEvent(event)
+	if err != nil {
+		return nil, err
+	}
+	return event, err
+}
+
+func marshalJwk(certASN1 []byte, certType cert.CertificateType) (map[string]interface{}, error) {
```

Comment:
Can't the JWK package do this for you given a x509 cert?

Reply:
Just checked, couldn't find it...